### PR TITLE
Typehintings and phpdocs

### PR DESCRIPTION
**Pull request by massimilianobraglia**
Since a major release is in development, I have just added so many typehintings and phpdocs compliant with PHP7.1.

Have a look and tell me what you think :)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,6 +16,10 @@
     ],
     "require": {
         "php": "^7.1",
+        "ext-json": "*",
+        "ext-libxml": "*",
+        "ext-simplexml": "*",
+        "ext-dom": "*",
```
Comment (alekitto) [inline composer.json]: These should be in `require-dev` section, as should be installed only if used, but they are not required to use serializer (none of them)

File: src/Exception/SyntaxErrorException.php [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 class SyntaxErrorException extends \Exception
 {
-    public function __construct($original, $value, $position)
+    public function __construct(string $original, string $value, ?int $position)
```
Comment (alekitto) [inline src/Exception/SyntaxErrorException.php:7]: Ok for `?int` but a cast is needed in `Type/Parser/Parser.php:L101`

File: src/GenericSerializationVisitor.php [hunk 3]
```diff
@@ -14,42 +14,75 @@ class GenericSerializationVisitor extends AbstractVisitor
...
+    /**
+     * @var mixed
+     */
     private $data;
 
-    public function setNavigator(GraphNavigator $navigator = null)
+    /**
+     * {@inheritdoc}
+     */
+    public function setNavigator(GraphNavigator $navigator = null): void
```
Comment (alekitto) [inline src/GenericSerializationVisitor.php]: `?GraphNavigator`

File: tests/Serializer/GraphNavigatorTest.php [hunk 4]
```diff
@@ -55,13 +77,16 @@ public function testNavigatorPassesInstanceOnSerialization()
         $context->getVisitor()->willReturn($visitor = $this->prophesize(VisitorInterface::class));
         $visitor->visitObject(Argument::cetera())->willReturn();
 
-        $visitor->startVisiting(Argument::cetera())->willReturn();
+        $visitor->startVisiting(Argument::cetera())->shouldBeCalled();
```
Comment (alekitto) [inline tests/Serializer/GraphNavigatorTest.php:80]: This is not under test and is not the assertion of this test. Please revert the changes to `willReturn`

**Review by alekitto — CHANGES_REQUESTED** [review]
Just some little fixes, but LGTM

File: composer.json (hunk 1 above)
Comment (massimilianobraglia) [reply composer.json]: 👍

File: src/Exception/SyntaxErrorException.php (hunk 2 above)
Comment (massimilianobraglia) [reply src/Exception/SyntaxErrorException.php:7]: 👍

File: src/GenericSerializationVisitor.php (hunk 3 above)
Comment (massimilianobraglia) [reply src/GenericSerializationVisitor.php]: 👍

File: tests/Serializer/GraphNavigatorTest.php (hunk 4 above)
Comment (massimilianobraglia) [reply tests/Serializer/GraphNavigatorTest.php:80]: 👍
